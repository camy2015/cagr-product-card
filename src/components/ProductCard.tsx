import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import React, { CSSProperties, createContext } from 'react';
import {
  InitialValues,
  Product,
  ProductCardHandlers,
  ProductContextProps,
  onChangeArgs,
} from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

interface Props {
  product: Product;
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    increaseBy,
    counter,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
