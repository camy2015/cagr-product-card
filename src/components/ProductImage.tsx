import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const { img = '' } = product;

  return (
    <img
      src={img ? img : noImage}
      alt="Product"
      className={`${styles.productImg} ${className}`}
      style={style}
    />
  );
};
