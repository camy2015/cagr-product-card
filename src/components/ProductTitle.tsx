import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const { id, title } = product;

  return (
    <>
      <span
        className={`${styles.productDescription} ${className}`}
        style={style}
      >
        {title}
      </span>
      <span
        className={`${styles.productDescription} ${className}`}
        style={style}
      >
        {id}
      </span>
    </>
  );
};
