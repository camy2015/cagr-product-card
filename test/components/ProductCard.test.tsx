import { describe, test, expect } from '@jest/globals';
import renderer, { act } from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product2 } from '../data/products';
import React from 'react';

describe('Tests in <ProductCard/>', () => {
  test('should display the component correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {() => <h1>Product Card</h1>}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should increase the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {({ count, increaseBy }) => (
          <>
            <h1>Product Card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}></button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
