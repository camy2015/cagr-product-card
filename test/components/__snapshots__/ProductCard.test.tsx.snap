// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <ProductCard/> should display the component correctly 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;

exports[`Tests in <ProductCard/> should increase the counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;
